### ISSUE TITLE
Why does the resnet50 model in torchvision output a different vector each time?

### ISSUE THREAD

**Issue by ponponon** [pytorch/vision#7937]
### 🐛 Describe the bug

```python
import torch
import torchvision
import torchvision.transforms as transforms
from PIL import Image
from torch import Tensor
import torch.nn as nn

model = torchvision.models.resnet50(weights='ResNet50_Weights.DEFAULT')
model.fc = nn.Linear(2048, 512)
model.eval()

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

image = Image.open('std.jpg')
image = transform(image).unsqueeze(0) 


with torch.no_grad():
    features: Tensor = model(image)


features_list = features.squeeze(0).tolist()

print(features_list[:10])
```

It's different every time you run it, and it's very, very different.

```shell
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[-0.29179805517196655, 0.2904847264289856, 0.12426053732633591, 0.02481590211391449, 0.329562783241272, 0.22396402060985565, -0.06559790670871735, -0.20587551593780518, 0.25109758973121643, -0.023244917392730713]
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[0.11942902207374573, 0.1826518476009369, -0.2526671588420868, -0.004894621670246124, -0.17024371027946472, -0.08633725345134735, 0.5060751438140869, 0.07067155838012695, 0.14896635711193085, 0.061379216611385345]
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[0.20133033394813538, -0.4602600336074829, 0.13579875230789185, 0.02763177454471588, 0.05834684893488884, 0.006434973329305649, 0.030948840081691742, -0.2761097848415375, 0.04298316687345505, 0.034981779754161835]
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[0.15419462323188782, -0.057300686836242676, 0.153386652469635, 0.19760870933532715, 0.2271871566772461, -0.15803731977939606, 0.14448338747024536, -0.0767395868897438, -0.01838969811797142, -0.033301010727882385]
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[0.18995943665504456, 0.16650597751140594, -0.1821695864200592, -0.13390550017356873, -0.3528384268283844, -0.10662798583507538, -0.06500241160392761, -0.32258665561676025, 0.22184018790721893, -0.1533258706331253]
(image2vector) ╭─ponponon@MBP13ARM ~/Desktop/code/me/resnet_example  ‹master*› 
╰─➤  python -u "/Users/ponponon/Desktop/code/me/resnet_example/resnet50_handle_image_into_vector.py"
[-0.3856864273548126, 0.11389009654521942, 0.11823548376560211, -0.1116785854101181, 0.24777428805828094, 0.45221805572509766, -0.1880977898836136, 0.07203484326601028, 0.29017287492752075, 0.05842892453074455]
```

There's no pattern to it?

I've obviously pre-trained the weights `weights='ResNet50_Weights.DEFAULT'`

### Versions

any version

**Comment by NicolasHug** [comment]
> model.fc = nn.Linear(2048, 512)


This is randomly initialized, that's why. If you want consistent results use `torch.manual_seed(some_int)`

**Comment by ponponon** [comment]
> > model.fc = nn.Linear(2048, 512)
> 
> This is randomly initialized, that's why. If you want consistent results use `torch.manual_seed(some_int)`

Are there any side effects? Like making reasoning slow?

**Comment by NicolasHug** [comment]
Hi @ponponon no, the random initialization doesn't affect performance whatsoever.

(I'll close this issue, since I believe the original concern was addressed)

**Comment by NicolasHug** [comment]
There is a lot to be said about this topic but the summary is that *not* setting a fixed random seed and getting random results is often desirable. In general, at least for research purposes, "seed engineering" (i.e. trying to find a good random seed) isn't exactly a good idea. You'd be more evaluating the performance of an initial state rather than the performance of a training procedure.

When you set a random seed, you either can get very lucky and get good results, or get unlucky and get bad results. Or get average luck and get average results. Pytorch doesn't want to make that decision on your behalf and by default provides maximum entropy i.e. random non-reproducible results. This is similar to how RNG in other packages (e.g. numpy) works.

**Comment by ponponon** [comment]
> There is a lot to be said about this topic but the summary is that _not_ setting a fixed random seed and getting random results is often desirable. In general, at least for research purposes, "seed engineering" (i.e. trying to find a good random seed) isn't exactly a good idea. You'd be more evaluating the performance of an initial state rather than the performance of a training procedure.关于这个主题有很多话要说，但总结是，不设置固定的随机种子并获得随机结果通常是可取的。一般来说，至少出于研究目的，“种子工程”（即试图找到一个好的随机种子）并不是一个好主意。您会更多地评估初始状态的性能而不是训练过程的性​​能。
> 
> When you set a random seed, you either can get very lucky and get good results, or get unlucky and get bad results. Or get average luck and get average results. Pytorch doesn't want to make that decision on your behalf and by default provides maximum entropy i.e. random non-reproducible results. This is similar to how RNG in other packages (e.g. numpy) works.当您设置随机种子时，您要么非常幸运并获得良好的结果，要么不幸并获得糟糕的结果。或者获得平均的运气并获得平均的结果。 Pytorch 不想代表您做出该决定，并且默认情况下提供最大熵，即随机的不可重现的结果。这与其他包（例如 numpy）中的 RNG 的工作原理类似。

model.fc = nn.Linear(2048, 512)

When I don't overwrite the existing fc with aspect statements, I get a stable 1000 dimensional output, the reason is because model = torchvision.models.resnet50(weights='ResNet50_Weights.DEFAULT') already contains the contents of the pre-trained fc layer right? I had thought that fc was simply modifying the dimensions, but now it seems that the fc layer also contains the weights parameter. It would be the difference in the weight parameters that would cause the fc to output different values right?

**Comment by NicolasHug** [comment]
yes, you're correct
